usr/sbin/p11sak/p11sak.c

```diff
@@ -3926,6 +4021,72 @@ static void print_utf8_attr(const char *attr, const CK_ATTRIBUTE *val,
     }
 }
 
+static void print_java_midp_secdom_attr(const char *attr, const CK_ATTRIBUTE *val,
+                                        int indent, bool sensitive)
+{
+    CK_JAVA_MIDP_SECURITY_DOMAIN secdom;
+
+    if (sensitive) {
+        printf("%*s%s: [sensitive]\n", indent, "", attr);
+        return;
+    } else if (val->ulValueLen == 0) {
+        printf("%*s%s: [no value]\n", indent, "", attr);
+        return;
+    }
+
+    secdom = *(CK_JAVA_MIDP_SECURITY_DOMAIN *)(val->pValue);
+
+    switch (secdom) {
+    case CK_SECURITY_DOMAIN_UNSPECIFIED:
+        printf("%*s%s: %s\n", indent, "", attr, "CK_SECURITY_DOMAIN_UNSPECIFIED");
+        break;
+    case CK_SECURITY_DOMAIN_MANUFACTURER:
+        printf("%*s%s: %s\n", indent, "", attr, "CK_SECURITY_DOMAIN_MANUFACTURER");
+        break;
+    case CK_SECURITY_DOMAIN_OPERATOR:
+        printf("%*s%s: %s\n", indent, "", attr, "CK_SECURITY_DOMAIN_OPERATOR");
+        break;
+    case CK_SECURITY_DOMAIN_THIRD_PARTY:
+        printf("%*s%s: %s\n", indent, "", attr, "CK_SECURITY_DOMAIN_THIRD_PARTY");
+        break;
+    default:
+        break;
+    }
+}
+
+static void print_cert_category_attr(const char *attr, const CK_ATTRIBUTE *val,
+                                     int indent, bool sensitive)
+{
+    CK_CERTIFICATE_CATEGORY category;
+
+    if (sensitive) {
+        printf("%*s%s: [sensitive]\n", indent, "", attr);
+        return;
+    } else if (val->ulValueLen == 0) {
+        printf("%*s%s: [no value]\n", indent, "", attr);
+        return;
+    }
+
+    category = *(CK_CERTIFICATE_CATEGORY *)(val->pValue);
+
+    switch (category) {
+    case CK_CERTIFICATE_CATEGORY_UNSPECIFIED:
+        printf("%*s%s: %s\n", indent, "", attr, "CK_CERTIFICATE_CATEGORY_UNSPECIFIED");
+        break;
+    case CK_CERTIFICATE_CATEGORY_TOKEN_USER:
+        printf("%*s%s: %s\n", indent, "", attr, "CK_CERTIFICATE_CATEGORY_TOKEN_USER");
+        break;
+    case CK_CERTIFICATE_CATEGORY_AUTHORITY:
+        printf("%*s%s: %s\n", indent, "", attr, "CK_CERTIFICATE_CATEGORY_AUTHORITY");
+        break;
+    case CK_CERTIFICATE_CATEGORY_OTHER_ENTITY:
+        printf("%*s%s: %s\n", indent, "", attr, "CK_CERTIFICATE_CATEGORY_OTHER_ENTITY");
+        break;
+    default:
+        break;
```

Comment:
Print the numeric value in hex for the default case:
```
        printf("%*s%s: 0x%lX\n", indent, "", attr,
               *(CK_CERTIFICATE_CATEGORY *)(val->pValue));
```

Reply:
Done.